txstratum/manager.py

```diff
@@ -133,8 +133,17 @@ def submit_solution(self, protocol: StratumProtocol, job: MinerJob, nonce: bytes
             assert isinstance(job, MinerTxJob)
             # Remove from queue.
             tx_job = job.tx_job
-            tx_job2 = self.tx_queue.popleft()
-            assert tx_job is tx_job2
+            self.log.info(f"Solution submitted for TxJob {tx_job}")
+            try:
+                self.tx_queue.remove(tx_job)
+            except ValueError:
+                # This means another miner submitted a solution shortly before
+                #  and so the txJob has been already removed from the queue
+                self.log.warning((
+                    "Tried removing tx_job from the queue but the queue does not contain this tx_job. "
+                    "Probably because 2 miners found a solution for it at the same time."
+                ))
```

Comment:
I think there is either a small misunderstanding or wrong (IMO) implementation. A job is supposed to be unique to a miner (if this is not the case the implementation is wrong IMO), as such it would be impossible for a different miner to submit a solution to the same job.

However it _is_ possible for a miner to submit more than one solution to the same job, in the context of the stratum protocol that makes sense, more than one `nonce` can solve the same job and if the miner has not switched jobs it will keep searching or another `nonce` on the same job, so that may actually be what is happening here. If it is implemented how I think it is, then the the "fix" IMO should be to either accept multiple solutions, or ignore it "cleanly", so maybe not with a warning, since this is expected.

Reply:
> as such it would be impossible for a different miner to submit a solution to the same job.

I've run a test to validate this, adding a log to show which miner solved which job. Apparently different miners are indeed submitting solutions to the same job.

Those are the logs I got:

```
txstratum.manager - INFO - 2021-12-01 11:46.36 [info     ] Solution submitted for TxJob <txstratum.jobs.TxJob object at 0x7f17be5f0340> by miner 39a1cb92-c919-4676-88d2-e8c2ae28a9fd
txstratum.manager - INFO - 2021-12-01 11:46.36 [info     ] TxJob solved                   propagate=False tx_job={'job_id': 'e3378b40fae08f7a979d211907ecb86b619ee49379b8f304d6368c4c41da4128', 'status': 'done', 'message': '', 'created_at': 1638369994.900104, 'tx': {'nonce': '000eaeba', 'parents': ['000026147de2f75ff1d645e28f52a1391830b42fc7d9894c594c8972d5b2a4d3', '000000e0f6eb02a681596c8add13efea019bb508856e82494785882309d1d10d'], 'timestamp': 1638369995, 'weight': 19.380205270799852}, 'timeout': 20.0, 'submitted_at': 1638369996.085033, 'total_time': 1.1849288940429688, 'expected_queue_time': 0, 'expected_mining_time': 0.10115834504104501, 'expected_total_time': 0.10115834504104501}
txstratum.manager - INFO - 2021-12-01 11:46.36 [info     ] Solution submitted for TxJob <txstratum.jobs.TxJob object at 0x7f17be5f0340> by miner c1d89a9f-11c2-4098-8262-2b0be621325c
txstratum.manager - WARNING - 2021-12-01 11:46.36 [warning  ] Tried removing tx_job from the queue but the queue does not contain this tx_job. Probably because 2 miners found a solution for it at the same time.
```

You can see that we have `Solution submitted for TxJob` 2 times for the same job, from different miners. Then the warning is logged.

> A job is supposed to be unique to a miner

I think this is true for the `MinerTxJob` instances. But they are wrappers for `TxJob` instances, and although the `MinerTxJob` instances are unique to each miner, the `TxJob` instances are not. And what we have in the queue are `TxJob` instances. That's why this behavior is happening, if I understood it correctly.

Reply:
Got it, that makes sense. Still, if this is expected and common, maybe it shouldn't be a warning. The only reason I think it makes sense for it to be a warning is because currently we can't distinguish it from a job that never existed.